libsycl/src/CMakeLists.txt

```
@@ -0,0 +1,94 @@
+include(WarningFlags)
+
+function(add_sycl_rt_library LIB_TARGET_NAME LIB_OBJ_NAME LIB_OUTPUT_NAME)
+  if (NOT LLVM_ENABLE_PIC)
+    message( FATAL_ERROR "Position-Independent Code generation is required for libsycl shared library" )
+  endif()
+
+  cmake_parse_arguments(ARG "" "" "COMPILE_OPTIONS;SOURCES" ${ARGN})
+
+  add_library(${LIB_OBJ_NAME} OBJECT ${ARG_SOURCES})
+
+  # Common compilation step setup
+  target_compile_definitions(${LIB_OBJ_NAME} PRIVATE
+              $<$<STREQUAL:${CMAKE_SYSTEM_NAME},Windows>:_LIBSYCL_BUILDING_LIBRARY>)
+  cxx_add_warning_flags(${LIB_OBJ_NAME} ${LIBSYCL_ENABLE_WERROR} ${LIBSYCL_ENABLE_PEDANTIC})
+
+  target_include_directories(
+    ${LIB_OBJ_NAME}
+    PRIVATE
+      ${CMAKE_CURRENT_SOURCE_DIR}
+      ${LIBSYCL_BUILD_INCLUDE_DIR}
+  )
+
+  add_library(${LIB_TARGET_NAME} SHARED
+              $<TARGET_OBJECTS:${LIB_OBJ_NAME}>)
+
+  add_dependencies(${LIB_OBJ_NAME}
+    sycl-headers
+  )
+
+  set_target_properties(${LIB_TARGET_NAME} PROPERTIES LINKER_LANGUAGE CXX)
+
+  if (CMAKE_SYSTEM_NAME STREQUAL Windows)
+    # Install stripped PDB
+    set(PDB_FILENAME "${LIB_TARGET_NAME}$<$<CONFIG:Debug>:d>")
+    check_linker_flag(CXX "LINKER:/PDBSTRIPPED:${PDB_FILENAME}.stripped.pdb"
+                          LINKER_SUPPORTS_PDBSTRIPPED)
+    if(LINKER_SUPPORTS_PDBSTRIPPED)
+      target_link_options(${LIB_TARGET_NAME}
+                          PRIVATE "LINKER:/PDBSTRIPPED:${PDB_FILENAME}.stripped.pdb")
+      install(FILES "${CMAKE_CURRENT_BINARY_DIR}/${PDB_FILENAME}.stripped.pdb"
+              DESTINATION ${CMAKE_INSTALL_PREFIX}/bin
+              RENAME "${PDB_FILENAME}.pdb"
+              COMPONENT ${LIB_TARGET_NAME}
+              OPTIONAL)
+    endif()
+
+    target_compile_options(${LIB_OBJ_NAME} PUBLIC /EHsc)
+  else()
+    target_compile_options(${LIB_OBJ_NAME} PUBLIC
+                           -fvisibility=hidden -fvisibility-inlines-hidden)
+
+    if (UNIX AND NOT APPLE)
+      set(linker_script "${CMAKE_CURRENT_SOURCE_DIR}/ld-version-script.txt")
+      target_link_libraries(
+          ${LIB_TARGET_NAME} PRIVATE "-Wl,--version-script=${linker_script}")
+      set_target_properties(${LIB_TARGET_NAME} PROPERTIES LINK_DEPENDS ${linker_script})
+    endif()
+  endif()
+
+  find_package(Threads REQUIRED)
+
+  target_link_libraries(${LIB_TARGET_NAME}
+    PRIVATE
+      ${CMAKE_DL_LIBS}
+      ${CMAKE_THREAD_LIBS_INIT}
+  )
+
+  set_target_properties(${LIB_TARGET_NAME} PROPERTIES
+                        VERSION ${LIBSYCL_VERSION_STRING}
+                        SOVERSION ${LIBSYCL_VERSION_STRING})
+  set_target_properties(${LIB_TARGET_NAME} PROPERTIES OUTPUT_NAME ${LIB_OUTPUT_NAME})
+endfunction(add_sycl_rt_library)
+
+set(LIBSYCL_SOURCES
+    "platform.cpp"
+)
+
+set(LIB_NAME "sycl")
+set(LIB_OUTPUT_NAME "${LIB_NAME}")
+if (CMAKE_SYSTEM_NAME STREQUAL Windows)
+  if (CMAKE_BUILD_TYPE STREQUAL Debug OR CMAKE_MSVC_RUNTIME_LIBRARY STREQUAL MultiThreadedDebugDLL)
```

Do we need to check for "MultiThreadedDebugDLL" specifically? I think the libc++ logic looks appropriate to use here. See https://github.com/llvm/llvm-project/blob/617af3cc50c48ec71889b893dd2658764f19abce/libcxx/CMakeLists.txt#L677-L679.
```suggestion
  if ((NOT CMAKE_MSVC_RUNTIME_LIBRARY AND uppercase_CMAKE_BUILD_TYPE STREQUAL "DEBUG")
      OR (CMAKE_MSVC_RUNTIME_LIBRARY MATCHES "Debug"))
```
I know we expect `CMAKE_MSVC_RUNTIME_LIBRARY` to always be set currently, but this logic will still work as expected if we relax that restriction later.

---

we don't expect it to be always set. We are ok with cmake default value MultiThreaded$<$<CONFIG:Debug>:Debug>DLL and handle explicitly set values.

---

MultiThreadedDebugDLL is the only option that is supported for sycl in Debug mode. Why should we introduce confusion by setting "CMAKE_MSVC_RUNTIME_LIBRARY MATCHES "Debug""?

---

Hmm, I didn't realize that `MultiThreadedDebugDLL` was the only supported option when building in debug mode. It seems that should be asserted rather than result in the "d" suffix being silently dropped. Should we also ensure that a MT variant is used? How about this?
```
  if (CMAKE_MSVC_RUNTIME_LIBRARY
      AND (NOT CMAKE_MSVC_RUNTIME_LIBRARY MATCHES "^MultiThreaded.*DLL$"))
    message(FATAL_ERROR "libsycl requires a multi-threaded DLL version of the MSVC CRT.")
  endif()
  if (uppercase_CMAKE_BUILD_TYPE STREQUAL "DEBUG")
    if (NOT CMAKE_MSVC_RUNTIME_LIBRARY STREQUAL "MultiThreadedDebugDLL")
      message(FATAL_ERROR "Debug builds of libsycl require a multi-threaded debug DLL version of the MSVC CRT.")
    endif()
    set(LIB_OUTPUT_NAME "${LIB_OUTPUT_NAME}d")
  endif()
```

---

"The intent is that the "d" suffix is appended if, and only if, the target is linked with a debug version of the MSVC CRT, correct?"
yes, it is how we do it now.
I would prefer to leave full name (MultiThreadedDebugDLL) since it represents better that SYCL RT is always a DLL, there is no other options.

---

The required use of a DLL is already assured at this point.

My concern is the "MultiThreaded" part and exact matching. Though CMake doesn't currently acknowledge it, there are additional MSVC CRT implementations; see the [MSVC `/Qspectre` option](https://learn.microsoft.com/en-us/cpp/build/reference/qspectre) and the special support for it that exists in MSVS for C++ projects via the ["Spectre Mitigation" property](https://learn.microsoft.com/en-us/cpp/build/reference/c-cpp-prop-page#spectre-mitigation). I think it isn't unreasonable that CMake might expose these additional run-time libraries; particularly for the MSBuild/Visual Studio generator. I don't know if the spectre CRT libraries are provided in debug form, but I would rather err on the side of adding the "d" suffix when "Debug" is present in the name than risk not adding it when libsycl is linked with an unknown CRT library.